File: leetcode/2608-shortest-cycle-in-a-graph/solution.py
from typing import List
from collections import deque


class Solution:
    def findShortestCycle(self, n: int, edges: List[List[int]]) -> int:
        graph = [[] for _ in range(n)]
        for a, b in edges:
            graph[a].append(b)
            graph[b].append(a)

        best = float('inf')

        # Run a BFS rooted at every node. Any edge found connecting two
        # already-visited nodes (other than the tree edge to the node's own
        # parent) closes a cycle whose length is the sum of both nodes'
        # BFS depths plus that one connecting edge.
        for start in range(n):
            dist = [-1] * n
            parent = [-1] * n
            dist[start] = 0
            queue = deque([start])

            while queue:
                u = queue.popleft()
                for v in graph[u]:
                    if dist[v] == -1:
                        dist[v] = dist[u] + 1
                        parent[v] = u
                        queue.append(v)
                    elif v != parent[u]:
                        best = min(best, dist[u] + dist[v] + 1)

        return -1 if best == float('inf') else best

File: leetcode/2608-shortest-cycle-in-a-graph/solution.cpp
#include <vector>
#include <queue>
#include <climits>
#include <algorithm>
using namespace std;

class Solution {
public:
    int findShortestCycle(int n, vector<vector<int>>& edges) {
        vector<vector<int>> graph(n);
        for (auto& e : edges) {
            graph[e[0]].push_back(e[1]);
            graph[e[1]].push_back(e[0]);
        }

        int best = INT_MAX;

        for (int start = 0; start < n; start++) {
            vector<int> dist(n, -1), parent(n, -1);
            dist[start] = 0;
            queue<int> q;
            q.push(start);

            while (!q.empty()) {
                int u = q.front(); q.pop();
                for (int v : graph[u]) {
                    if (dist[v] == -1) {
                        dist[v] = dist[u] + 1;
                        parent[v] = u;
                        q.push(v);
                    } else if (v != parent[u]) {
                        best = min(best, dist[u] + dist[v] + 1);
                    }
                }
            }
        }

        return best == INT_MAX ? -1 : best;
    }
};
